pkg/kubelet/dockertools/manager.go

```diff
@@ -1791,6 +1865,43 @@ func (dm *DockerManager) SyncPod(pod *api.Pod, _ api.PodStatus, podStatus *kubec
 		}
 	}
 
+	// only the last execution of an init container should be preserved, and only preserve it if it is in the
+	// list of init containers to keep.
+	initContainerNames := sets.NewString()
+	for _, container := range pod.Spec.InitContainers {
+		initContainerNames.Insert(container.Name)
+	}
+	for name := range initContainerNames {
+		count := 0
+		for _, status := range podStatus.ContainerStatuses {
+			if !initContainerNames.Has(status.Name) || status.State != kubecontainer.ContainerStateExited {
+				continue
+			}
+			count++
+			if count == 1 {
+				continue
+			}
+			if _, ok := containerChanges.InitContainersToKeep[kubecontainer.DockerID(status.ID.ID)]; ok {
+				continue
+			}
+			glog.V(2).Infof("Removing init container %q instance %q %d", status.Name, status.ID.ID, count)
+			if err := dm.client.RemoveContainer(docker.RemoveContainerOptions{ID: status.ID.ID, RemoveVolumes: true}); err != nil {
+				if _, ok := err.(*docker.NoSuchContainer); ok {
```

Comment:
I need to DRY up this and the container init (as your next comment
highlights).

On Fri, Apr 8, 2016 at 4:41 PM, Prashanth B <notifications@github.com>
wrote:

> In pkg/kubelet/dockertools/manager.go
> <https://github.com/kubernetes/kubernetes/pull/23567#discussion_r59086260>
> :
>
> > +	for name := range initContainerNames {
> > +		count := 0
> > +		for _, status := range podStatus.ContainerStatuses {
> > +			if !initContainerNames.Has(status.Name) || status.State != kubecontainer.ContainerStateExited {
> > +				continue
> > +			}
> > +			count++
> > +			if count == 1 {
> > +				continue
> > +			}
> > +			if _, ok := containerChanges.InitContainersToKeep[kubecontainer.DockerID(status.ID.ID)]; ok {
> > +				continue
> > +			}
> > +			glog.V(2).Infof("Removing init container %q instance %q %d", status.Name, status.ID.ID, count)
> > +			if err := dm.client.RemoveContainer(docker.RemoveContainerOptions{ID: status.ID.ID, RemoveVolumes: true}); err != nil {
> > +				if _, ok := err.(*docker.NoSuchContainer); ok {
>
> nit: I'd write a function that just returns the one true status and a list
> of toDelete, and remove all the toDeletes, instead of the counter
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/23567/files/21f0e955e3f1d05950fed76280c28d3e81d90742..ad7a96fe3b9dda804fee77e41f7e14ad0b4513d7#r59086260>
>